revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,208 @@
+				defer wg.Done()
+				certResults[i] = certOCSPStatus(cert, opts.CertChain[i+1], opts)
+			}(i, cert)
+		}
+		// Last is root cert, which will never be revoked by OCSP
+		certResults[len(opts.CertChain)-1] = []error{nil}
+
+		wg.Wait()
+	}
+	return certResults
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, opts Options) []error {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return []error{NoOCSPServerError{}}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		if !strings.HasPrefix(strings.ToLower(server), "http://") {
+			// This function is only able to check servers that are accessible via HTTP
+			serverErrs[serverIndex] = CheckOCSPError{Err: errors.New("OCSPServer must be accessible over HTTP")}
+			continue
+		}
+		resp, err := ocspRequest(cert, issuer, server, opts)
+		if err != nil {
+			// If there is a server error, attempt all servers before determining what to return
+			// to the user
+			if errors.Is(err, TimeoutError{}) {
+				serverErrs[serverIndex] = err
+			} else {
+				serverErrs[serverIndex] = CheckOCSPError{Err: err}
+			}
+			continue
+		}
+
+		if time.Now().After(resp.NextUpdate) {
+			serverErrs[serverIndex] = CheckOCSPError{Err: errors.New("expired OCSP response")}
+			continue
+		}
+
+		foundNoCheck := false
+		for _, extension := range resp.Extensions {
+			if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+				foundNoCheck = true
+				break
+			}
+		}
+		if !foundNoCheck {
+			// This will be ignored until CRL is implemented
+			// If it isn't found, CRL should be used to verify the OCSP response
+			// TODO: add CRL support
+			// https://github.com/notaryproject/notation-core-go/issues/125
+			serverErrs[serverIndex] = PKIXNoCheckError{}
+			continue
+		}
+
+		switch resp.Status {
+		case ocsp.Revoked:
+			if opts.SigningTime.After(resp.RevokedAt) {
+				return []error{RevokedError{}}
+			} else {
+				return []error{nil}
+			}
+		case ocsp.Good:
+			return []error{nil}
+		default:
+			return []error{UnknownStatusError{}}
+		}
+
+	}
+	return serverErrs
+}
+
+func ocspRequest(cert, issuer *x509.Certificate, server string, opts Options) (*ocsp.Response, error) {
+	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{Hash: crypto.SHA256})
+	if err != nil {
+		return nil, err
+	}
+
+	var resp *http.Response
+	encodedReq := base64.StdEncoding.EncodeToString(ocspRequest)
+	if len(encodedReq) >= 255 {
```

Comment:
Please use `base64.URLEncoding.EncodedLen()` to determine the length for performance considerations.